symbolic-debuginfo/src/macho/bcsymbolmap.rs

```diff
@@ -187,6 +191,194 @@ impl<'a, 'd> Iterator for BcSymbolMapIterator<'a, 'd> {
+            dsym_uuid,
+        })
+    }
+
+    /// Parses a PropertyList containing a `DBGOriginalUUID` mapping.
+    ///
+    /// This is a convenience version of [`BitcodeUuidMapping::parse_plist`] which extracts
+    /// the UUID from the `filename`.
+    ///
+    /// The `filename` may contain multiple path segments, the stem of the filename segment
+    /// should contain the UUID of the `dSYM`.  This is the format the PList is normally
+    /// found in a `dSYM` directory structure.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::path::Path;
+    /// use symbolic_common::DebugId;
+    /// use symbolic_debuginfo::macho::BitcodeUuidMapping;
+    ///
+    /// let filename = Path::new("Contents/Resources/2D10C42F-591D-3265-B147-78BA0868073F.plist");
+    /// # let filename = Path::new("tests/fixtures/2d10c42f-591d-3265-b147-78ba0868073f.plist");
+    /// let data = std::fs::read(filename).unwrap();
+    ///
+    /// let uuid_map = BitcodeUuidMapping::parse_plist_with_filename(filename, &data).unwrap();
+    ///
+    /// assert_eq!(
+    ///     uuid_map.dsym_uuid(),
+    ///     "2d10c42f-591d-3265-b147-78ba0868073f".parse().unwrap(),
+    /// );
+    /// assert_eq!(
+    ///     uuid_map.bitcode_uuid(),
+    ///     "c8374b6d-6e96-34d8-ae38-efaa5fec424f".parse().unwrap(),
+    /// )
+    /// ```
+    pub fn parse_plist_with_filename(filename: &Path, data: &[u8]) -> Result<Self, PListError> {
+        let dsym_uuid = filename
+            .file_stem()
+            .ok_or_else(|| PListError::from(PListErrorKind::ParseFilename))?
+            .to_str()
+            .ok_or_else(|| PListError::from(PListErrorKind::ParseFilename))?
+            .parse()?;
+        Self::parse_plist(dsym_uuid, data)
+    }
+
+    /// Returns the UUID of the original bitcode.
+    pub fn bitcode_uuid(&self) -> DebugId {
+        self.bitcode_uuid
+    }
+
+    /// Returns the UUID of the compiled binary and associated `dSYM`.
+    pub fn dsym_uuid(&self) -> DebugId {
+        self.dsym_uuid
+    }
+}
+
+fn uuid_from_plist(data: &[u8]) -> Result<DebugId, PListError> {
+    let plist = Element::from_reader(Cursor::new(data))?;
+
+    let version = plist
+        .get_attr("version")
+        .ok_or_else(|| PListError::from(PListErrorKind::Schema))?;
+    if version != "1.0" {
+        return Err(PListError::from(PListErrorKind::Schema));
+    }
+    let dict = plist
+        .find("dict")
+        .ok_or_else(|| PListError::from(PListErrorKind::Schema))?;
+
+    let mut found_key = false;
+    let mut raw_original = None;
+    for element in dict.children() {
+        if element.tag().name() == "key" && element.text() == "DBGOriginalUUID" {
+            found_key = true;
+        } else if found_key {
+            raw_original = Some(element.text().to_string());
+            break;
+        }
+    }
+
+    match raw_original {
+        Some(raw) => raw.parse().map_err(Into::into),
+        None => Err(PListError::from(PListErrorKind::Schema)),
```

Comment:
Just a stylistic suggestion:
You can define an inner fn that returns `Option<&str>`, use short-circuit `?` inside, and then call that with one `ok_or_else` instead of doing so in basically 3-4 places above.